### ISSUE TITLE
Java false positive: XXE via XMLInputFactory

### ISSUE THREAD

**Issue by vlkl-sap** [github/codeql#7199]
The following Java code snippet triggers a false positive XXE alert:
```
var xif = XMLInputFactory.newInstance();
xif.setProperty(XMLConstants.ACCESS_EXTERNAL_DTD, "");
xif.setProperty(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
var xsr = xif.createXMLStreamReader(new StringReader(xml));
```

**Comment by smowton** [comment]
I think ACCESS_EXTERNAL_STYLESHEET needs setting too to be safe? I've made a PR doing that at https://github.com/github/codeql/pull/7220

**Comment by m-y-mo** [comment]
Thanks for raising this issue. The current XXE query is based on the guidelines set out by OWASP: https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html#xmlinputfactory-a-stax-parser
As we are not confident about the safety of deviating from the OWASP guidelines, (e.g. edge cases or older versions of Java that may lead to bypass), we've decided not to change the query. As the query is open source, users can modify it according to their specific circumstances. We'll keep the PR #7220 so that any user may use it to modify the query as they see fit. Thank you very much.

**Comment by vlkl-sap** [comment]
Thanks for taking a look. I understand what you are saying and I do sympathize, but I cannot let you off the hook that easily.

1) Everything you said (I think) applies to the SECURE_PROCESSING sanitizer, which you currently do recognize.

2) If you decide not to recognize a particular sanitizer, then I think that decision must be documented in the help text for the query. Otherwise, the customer is left guessing.

3) I understand that XXE is a minefield with implementation-specific behaviors and absolutely terrible vendor documentation. Ultimately, someone would have to do the empirical legwork, dive into the code, and get a reasonable picture of which mitigation works where - and give the customer a way to systematically ignore the legacy resp. edge cases, if they wish.

In other words, there is a need for action in this general area. Thanks for listening to my TED talk.

**Comment by m-y-mo** [comment]
Thanks for your reply. I agree that the current complication in the XXE query and the different configurations that it considers as safe may be confusing to users. We'll change the documentation and state clearly that we only consider the configurations set out in the OWASP guideline as safe so that it is clear to users what is being used. Thank you very much for your helpful suggestions and feedback. We'll prepare a PR for the change of the documentation and let you know. Please feel free to leave comment and give us your feedback on the PR. Thanks!

**Comment by smowton** [comment]
@m-y-mo @vlkl-sap how about https://github.com/github/codeql/pull/7229

**Comment by m-y-mo** [comment]
@vlkl-sap Thanks for the feedback. We've reviewed the use of SECURE_PROCESSING and concluded that it is an insufficient defence against XXE on its own. We therefore decided to remove SECURE_PROCESSING as a standalone sanitiser against XXE and follow the OWASP recommendations. Thank you very much for making us aware of this and help us to improve the query. Please let us know if you have any other feedbacks. Thanks!

**Comment by smowton** [comment]
Both PRs are merged, so marking issue resolved by making a uniform policy in favour of the OWASP recommended XXE defence for a given parser.
